src/actions/sale/purchaseDomains.ts

```diff
@@ -0,0 +1,166 @@
+import * as ethers from "ethers";
+import { getSaleStatus } from ".";
+import { Sale, WapeSale } from "../../contracts/types";
+import { isTransitionToPublicPhasePending } from "../../helpers";
+import { Claim, SaleStatus, Mintlist, Maybe, SalePhase } from "../../types";
+
+const abi = ["function masterCopy() external view returns (address)"];
+
+const errorCheck = async (condition: boolean, errorMessage: string) => {
+  if (condition) {
+    throw errorMessage;
+  }
+};
+
+const generateAccessList = (
+  userAddress: string,
+  gnosisSafeProxyAddress: string,
+  gnosisSafeImplAddress: string
+) => {
+  return [
+    {
+      address: gnosisSafeProxyAddress,
+      storageKeys: [
+        "0x0000000000000000000000000000000000000000000000000000000000000000",
+      ],
+    },
+    {
+      address: gnosisSafeImplAddress,
+      storageKeys: [
+        "0x0000000000000000000000000000000000000000000000000000000000000000",
+      ],
+    },
+    {
+      address: userAddress,
+      storageKeys: [],
+    },
+  ];
+};
+
+export const purchaseDomains = async (
+  count: ethers.BigNumber,
+  signer: ethers.Signer,
+  contract: Sale,
+  mintlist: Mintlist,
+): Promise<ethers.ContractTransaction> => {
+  const status: SalePhase =
+    (await contract.salePhase()) as SalePhase;
+
+  errorCheck(
+    status === SalePhase.Inactive || status === SalePhase.ReadyForNewSale,
+    "Cannot purchase domains: Sale prepared but not yet started"
+  );
+
+  errorCheck(status === SalePhase.Inactive, "Cannot purchase domains: No sale in progress");
```

Comment:
will this ever get hit if a sale is Inactive?

Reply:
good catch. I want distinct errors for each phase, but I'm checking for either invalid phase above.